### DISCUSSION TITLE
Importing Server Action to Client Component - is this a bug?

### DISCUSSION THREAD

**Discussion by lucaspiresfernandes** [vercel/next.js#49672] [App Router]
Hi there. Was playing with Server Actions and found something interesting. I don't know if I'm just being stupid or if this is an actual bug, so hear me out.

My server action, called `createSession`, does exactly what it says: validates login information and creates a new user session if applicable. This action utilizes a database connection, a set of cookie utilities and jwt for cookie encryption.

Everything was working just fine, but then I imported `server-only` package to my dependencies listed above and now I'm getting a lot of errors from my `server-only` package. Which makes me wonder: if the dependencies are being used by the Server Action, which runs on the server, why the heck are they complaining about running on the client? Could it be a code-splitting issue?

For better context, here's my code:

page.tsx (stripped for better readability)
``` ts
'use client';

import { createSession } from '@/app/sign-in/actions';

const SignIn: React.FC = () => {
	const onSubmit: React.FormEventHandler<HTMLFormElement> = (ev) => {
		ev.preventDefault();
		const formData = new FormData(ev.currentTarget);
		createSession({
			email: formData.get('email') as string,
			password: formData.get('password') as string,
		});
	};

	return (
		<main>
			<h1>Login</h1>
			<form onSubmit={onSubmit}>{/* Login form stuff */}</form>
		</main>
	);
};

export default SignIn;
```

actions.ts
``` ts
'use server';

import { PASSWORD_POLICY } from '@/utils/account';
import { setSessionCookie } from '@/utils/cookie';
import prisma from '@/utils/db';

const Credentials = z.object({
	email: z.string().email(),
	password: z.string().min(PASSWORD_POLICY.minLength),
});

export type CreateSessionRequest= z.infer<typeof Credentials>;
type CreateSessionResponse = { success: true; id: number } | { success: false; reason: 'parse' | 'invalid' };

export async function createSession(req: CreateSessionRequest): Promise<CreateSessionResponse> {
	// Do session stuff.
}
```

libraries imported from `@/utils/*` folder are the ones marked with `import 'server-only'`.

How the hell is this code running on client-side? Is this a bug?

**Comment by icyJoseph** [comment]
Interesting, I guess it is not running on the client, but I wonder if because the idea with client only and server only packages, is to absolutely prevent accidents

- https://www.npmjs.com/package/client-only
- https://www.npmjs.com/package/server-only

Then I'd guess that the tooling hasn't yet been able to use the `'use server'` directive as a way to opt-out of the "client bundling" context?

AFAIK, the way modules are loaded can lead to unexpected bundles. Most recently I had an issue where a simple string import brought in browserify's crypto onto my client side. Similarly to how, we shouldn't blindly trust tree shaking, 

I guess, when you label something `server-only`, the chain of imports crashes out if it originates from a 'use client' component.

**Reply by lucaspiresfernandes** [reply]
Yeah, it's such a fine line between client and server that `server-only` can't really do its job properly, I guess.

**Reply by icyJoseph** [reply]
That's why it is an alpha release, good to point it out though.
